File: commons/Tree.py
from collections import deque
from typing import Optional


def serialize(root: 'TreeNode') -> list[Optional[int]]:
    queue = deque([root])
    result = []

    while queue:
        curr = queue.popleft()

        if not curr:
            result.append(None)
        else:
            result.append(curr.val)
            queue.append(curr.left)
            queue.append(curr.right)

    return result


def parse(serialized: list[Optional[int]]):
    if not serialized:
        return None

    values = deque(serialized)
    head = TreeNode(values.popleft())
    nodes = deque([head])

    while values:
        curr_node = nodes.popleft()

        left_val = values.popleft()
        if left_val is not None:
            curr_node.left = TreeNode(int(left_val))
            nodes.append(curr_node.left)

        right_val = values.popleft()
        if right_val is not None:
            curr_node.right = TreeNode(int(right_val))
            nodes.append(curr_node.right)

    return head


class TreeNode:
    def __init__(self, val: int, left:Optional['TreeNode']=None, right:Optional['TreeNode']=None) -> None:
        self.val = val
        self.left = left
        self.right = right


    def __str__(self):
        temp = ['null' if x is None else str(x) for x in serialize(self)]
        return ','.join(temp)


    @staticmethod
    def deserialize(serialized: str):
        return parse(serialized)

File: commons/Tree.ts
// from collections import deque
// from typing import Optional

const serialize = <T>(root: TreeNode<T> | null): Array<T | null> => {
  const queue = [root]
  const result = []

  while (queue) {
    const curr = queue.shift()

    if (!curr) {
      result.push(null)
    } else {
      result.push(curr.val)
      queue.push(curr.left)
      queue.push(curr.right)
    }
  }

  return result
}

const parse = <T>(serialized: Array<T | null>): TreeNode<T> | null => {
  if (!serialized.length) return null

  const values = [...serialized]
  const head = new TreeNode(values.shift()!)
  const nodes = [head]

  while (values.length) {
    const currNode = nodes.shift()!

    const leftVal = values.shift()
    if (leftVal !== null && typeof leftVal !== 'undefined') {
      currNode.left = new TreeNode(leftVal)
      nodes.push(currNode.left)
    }

    const rightVal = values.shift()
    if (rightVal !== null && typeof rightVal !== 'undefined') {
      currNode.right = new TreeNode(rightVal)
      nodes.push(currNode.right)
    }
  }

  return head
}

export class TreeNode<T = any> {
  static fromArray<T>(serialized: Array<T | null>) {
    return parse(serialized)
  }

  constructor(
    public val: T,
    public left: TreeNode | null = null,
    public right: TreeNode | null = null,
  ) {
  }

  toString(): string {
    return serialize(this).join(',')
  }
}
